ckanext/canada/plugins.py

```diff
@@ -437,3 +440,55 @@ def after_show(self, context, data_dict):
 
     def update_facet_titles(self, facet_titles):
         return facet_titles
+
+class CanadaActivity(p.SingletonPlugin):
+    p.implements(p.IActions)
+    p.implements(IActivity)
+
+    def get_actions(self):
+        #override datastore_upsert and datastore_delete
+        for plugin in p.PluginImplementations(p.IActions):
+            if plugin == self:
+                continue
+            actions = plugin.get_actions()
+            if 'datastore_upsert' in actions and (
+                    'datastore_delete' in actions):
+                old_upsert = actions['datastore_upsert']
+
+                def new_datastore_upsert(context, data_dict):
+                    act.datastore_activity_create(context, data_dict)
+                    return old_upsert(context, data_dict)
```

Comment:
The activity creation should be called after the original upsert/delete calls because those calls might raise a ValidationError.

As a smaller thing: wherever possible don't use (context, data_dict) as parameters to new functions. proper parameters are always easier to understand.